### DISCUSSION TITLE
inTransaction is not working when not using commit() for inital autoCommit = false connections;

### DISCUSSION THREAD

**Discussion by RezaLabudeBSM** [jdbi/jdbi#2662] [Q&A]
Hello,

and again another problem i do not understand why i can not get it to work correctly:
The Working Code which really deletes the rows in the database after the explicit commit:

`

	public int deleteTable(String sqlDeleteStatement, Map<String, Object> deleteBindMap) {
		return jdbi.inTransaction( transactionHandle -> {
				transactionHandle.begin();
				int count = transactionHandle.createUpdate(sqlDeleteStatement).bindMap(deleteBindMap).execute();
				transactionHandle.commit();
				return count;
			}
		);
	}

`
with this stacktrace:

![image](https://github.com/jdbi/jdbi/assets/72243112/53066d90-4509-439c-b498-1fe67ea946ce)

I then changed the code without the explicit begin and commit

`

	public int deleteTable(String sqlDeleteStatement, Map<String, Object> deleteBindMap) {
		return jdbi.inTransaction( transactionHandle -> {
				return transactionHandle.createUpdate(sqlDeleteStatement).bindMap(deleteBindMap).execute();
			}
		);
	}

`
But this time the delete is not commited in the database and therefore not persisted = the rows are still there and not deleted.
The stacktrace is the same and does also have the inTranscation

![image](https://github.com/jdbi/jdbi/assets/72243112/2b032737-2b40-41b4-9565-f669b7196efd)

So why is the second example not working. As far as i understood it is not necessary to use exlicit transaction commands begin and commit.
(I'm running in a spring boot container, but i did not use the TransactionAwareDataSourceProxy and want to rely on the JDBI transaction management (when i activate this Proxy, then it works correctly with the Spring @Transactional Annotation, but this is not what i want))

**Comment by RezaLabudeBSM** [comment]
After some debugging and experimenting, i think i found the problem:

`

        BoundLocalTransactionHandler(Handle handle) throws SQLException {
            this.initialAutocommit = handle.getConnection().getAutoCommit();
            this.handlerState = initialAutocommit ? State.OUTSIDE_TRANSACTION : State.AFTER_BEGIN;
        }

`
During this creation of a BoundLocalTransactionHandler the inital handlerState  depends on the state of the initialAutocommit. Since i normally set this to false, the state was for me always AFTER_BEGIN, even though no real start of an transaction was done. 
When then later this isInTransaction check happens:

`

   public <R, X extends Exception> R inTransaction(HandleCallback<R, X> callback) throws X {
        return isInTransaction()
                ? callback.withHandle(this)
                : transactionHandler.inTransaction(this, callback);
    }


`


it has already a started transaction and never gets to the transactionHandler.inTransaction which does the begin and commit as expected. So my question is why do you do this kind of initialisation? it looks like someone wanted to save thereby the call to the begin() method as the connection is already setAutoCommit(false) and therefore the handlerState can go directly to AFTER_BEGIN, but with this "optimisation" the inTransaction does not work correctly anymore. 
I hope my analysis is correct and you can fix this.

**Comment by RezaLabudeBSM** [comment]
I checked the code and this behaviour was introduced because of https://github.com/jdbi/jdbi/issues/2491. But with this inital setting of handlerState to AFTER_BEGIN there is no chance that the code inside of transactionHandler.inTransaction will be called as this check will return true because of the second false for getAutoCommit:
        
`

    public boolean isInTransaction(Handle handle) {
            try {
                return handlerState == State.IN_TRANSACTION || !handle.getConnection().getAutoCommit();
            } catch (SQLException e) {
                throw new TransactionException("Failed to test for transaction status", e);
            }
        }

`

The problem that there is no test that tests, whether a commit() is called, when autocommit() is set to false. 
I would therefore suggest to improve the tests in this way:

    @Test
    public void testCallback() {
        String woot = h.inTransaction(x -> "Woot!");
        assertThat(commit).isOne();
        assertThat(woot).isEqualTo("Woot!");
    }
    
    @Test
    public void testCallbackWithAutoCommitFalse() {
    	  Jdbi jdbi = Jdbi.create(() -> {
              // create connection with auto-commit == false
              Connection connection = DriverManager.getConnection(h2Extension.getUri());
              connection.setAutoCommit(false);
              return connection;
          });
    	  jdbi.setTransactionHandler(txSpy);
        String woot = jdbi.inTransaction(x -> "Woot!");
        assertThat(commit).isOne();
        assertThat(woot).isEqualTo("Woot!");
    }


`

I tested it with the actual implemention and now you have a failing test

**Comment by RezaLabudeBSM** [comment]
My suggestion to fix this problem would be to remove the || !handle.getConnection().getAutoCommit() as this code was from the previous implementation without the statemachine via the enum. Now you have this enum and you should trust its state to observe the state of the transaction correctly. Otherwise it will be difficult to solve this problem for both autoCommit states.

**Comment by stevenschlansker** [comment]
This is filed as issue #2663 - let's keep the conversation in one place.
